File: java/BSTest.java

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.file.FileSystems;
import java.nio.file.StandardOpenOption;


public class BSTest {
  static final String H1 = "**************************************************************************";
  static final String H2 = "__________________________________________________________________________";

  public static int pack4B(byte b1, byte b2, byte b3, byte b4) {
    // IMPORTANT: BigEndian by default
    ByteBuffer dbuf = ByteBuffer.allocate(4);
    dbuf.put(b4);
    dbuf.put(b3);
    dbuf.put(b2);
    dbuf.put(b1);
    return dbuf.getInt(0);
  }

  public static byte lookUpByte_0(ByteStream BS, int targetIndex) throws Exception {
    for (int i = 0; i < targetIndex; ++i) {
      BS.getByte();
    }
    return BS.getByte();
  }

  public static void main(String[] args) throws IOException, Exception {
    System.out.println(H1);
    System.out.println("Initializing ByteStream");
    ByteStream BS = new ByteStream();
    System.out.println(H1);

    System.out.println("putByte Test");
    System.out.println(H2);
    byte[] ptr;
    boolean verif;
    for (int i = 0; i < 10; i++) {
      ptr = BS.getByteStream();
      verif = true;
      while (verif) {
        BS.putByte((byte) 1); // [0 - 255], overflow not allowed
        verif = ptr == BS.getByteStream();
      }
      System.out.printf("old: %s\tnew: %s\tlength: %d\n", ptr,  BS.getByteStream(), BS.getByteStream().length);
    }
    ptr = BS.getByteStream();
    System.out.println(H2);
    System.out.printf("0: %d\t\t65536: %d\t\t524287: %d\t\t1000000: %d\n",
    ptr[0], ptr[65536], ptr[524287], ptr[1000000]);
    System.out.println(H1);
    /* **************************************************** */
    System.out.printf("%42s\n", "putBytes_0 Test");
    System.out.printf("%s\n", H2);
    BS = new ByteStream();
    byte[] arr = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
    for (int i = 0; i < 10; i++) {
      ptr = BS.getByteStream();
      verif = true;
      while (verif) {
        BS.putBytes(arr, 0, 16); // Whole array, sequences of [0-16]
        verif = ptr == BS.getByteStream();
      }
      System.out.printf("old: %s\tnew: %s\tlength: %d\n",
        ptr, BS.getByteStream(), BS.getByteStream().length);
    }
    System.out.printf("%s\n", H2);
    ptr = BS.getByteStream();
    System.out.printf("0: %d\t\t67541: %d\t\t540639: %d\t\t1000000: %d\n",
              ptr[0], ptr[67541], ptr[540639], ptr[1000000]);
    System.out.printf("%s\n", H1);
    /* **************************************************** */
    System.out.printf("%42s\n", "putBytes_1 Test");
    System.out.printf("%s\n", H2);
    BS = new ByteStream();
    for (int i = 0; i < 10; i++) {
      ptr = BS.getByteStream();
      verif = true;
      while (verif) {
        /*
         * Just to remind the byte format of Java (same as C's signed char)
         * UINT8    BYTE
         *   1       1
         *  127     127
         *  128    -128 (0 from another alternative dimension, somehow)
         *  129    -127
         *  255     -1
         */
        BS.putBytes(2147483647, 256); // Max 4B value
        verif = ptr == BS.getByteStream();
      }
      System.out.printf("old: %s  \tnew: %s  \tlength: %d\n",
        ptr, BS.getByteStream(), BS.getByteStream().length);
    }
    System.out.printf("%s\n", H2);
    ptr = BS.getByteStream();
    System.out.printf("0: %d\t\t65534: %d\t\t524287: %d\t\t1000000: %d\n",
              ptr[0], ptr[65534], ptr[524287], ptr[1000000]);
    System.out.printf("%s\n", H1);
    /* **************************************************** */
    BS.reset();
    System.out.printf("%36s\n", "getByte_1 Test (normal mode)");
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", BS.getByte(0), BS.getByteStream()[0]);
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", BS.getByte(65534), BS.getByteStream()[65534]);
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", BS.getByte(524287), BS.getByteStream()[524287]);
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 1000000\t\t Got: %d\t\t\tExpected: %d\n", BS.getByte(1000000), BS.getByteStream()[1000000]);

    System.out.printf("%s\n", H1);
    /* **************************************************** */
    System.out.printf("%36s\n", "getByte_0 Test (normal mode)");
    System.out.printf("%s\n", H2);

    // Just don't want employ resets here for future tests
    System.out.printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS, 0), BS.getByteStream()[0]);
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS, 65534 - 1), BS.getByteStream()[65534]);
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS, 524287 - 65535), BS.getByteStream()[524287]);
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 1000000\t\t Got: %d\t\t\tExpected: %d\n", lookUpByte_0(BS, 1000000 - 524288), BS.getByteStream()[1000000]);
    System.out.printf("%s\n", H2);
    /* **************************************************** */
    System.out.printf("%36s\n", "getBytes Test (normal mode)");
    System.out.printf("%s\n", H2);
    BS.reset(); // lol I lied

    System.out.printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", BS.getBytes(1), pack4B(BS.getByteStream()[0], (byte) 0, (byte) 0, (byte) 0));
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 1-2\t\t Got: %d\t\tExpected: %d\n", BS.getBytes(2), pack4B(BS.getByteStream()[2], BS.getByteStream()[1], (byte) 0, (byte) 0));
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 3-5\t\t Got: %d\t\tExpected: %d\n", BS.getBytes(3), pack4B(BS.getByteStream()[5], BS.getByteStream()[4], BS.getByteStream()[3], (byte) 0));
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 6-9\t\t Got: %d\t\tExpected: %d\n", BS.getBytes(4), pack4B(BS.getByteStream()[9], BS.getByteStream()[8], BS.getByteStream()[7], BS.getByteStream()[6]));
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 10-14\t\t Got: %d \t\tExpected: %d\n", BS.getBytes(5),
      pack4B(BS.getByteStream()[14], BS.getByteStream()[13], BS.getByteStream()[12], BS.getByteStream()[11]) + pack4B(BS.getByteStream()[10], (byte) 0, (byte) 0, (byte) 0));
    System.out.printf("%s\n", H2);

    System.out.printf("Index: 15-22\t\t Got: %d \tExpected: %d\n", BS.getBytes(8),
      pack4B(BS.getByteStream()[22], BS.getByteStream()[21], BS.getByteStream()[20], BS.getByteStream()[19])
      + pack4B(BS.getByteStream()[18], BS.getByteStream()[17], BS.getByteStream()[16], BS.getByteStream()[15]));
    System.out.printf("%s\n", H2);
    /* Code for displaying the bits of the output number step by step in function getBytes(BS, num)

        // Simplified status of the buffer
        System.out.printf("\t\t%d\t%d|%d|%d|%d\n", (char) getByte, (num >> 0) & 0xff, (num >> 8) & 0xff, (num >> 16) & 0xff, (num >> 24) & 0xff);

        // New byte added
        System.out.printf("(%d) -> %d = ", (char) getByte, b * 8);
        for (int j = 7; j >= 0; j--) {
          System.out.printf("%c", (((char) getByte) & (1 << j)) ? '1' : '0');
        }
        System.out.printf(" ");
        System.out.printf("\n");

        // Detailed status of the buffer
        System.out.printf("(%d) %*s = ", num, 8, "->");
        for (int i = 0; i <= 3; i++) {
          for (int j = 7; j >= 0; j--) {
            System.out.printf("%c", ((num >> (i * 8)) & (1 << j)) ? '1' : '0');
          }
          System.out.printf(" ");
        }
        System.out.printf("\n");
    */
    /* **************************************************** */
    System.out.printf("%36s\n", "write_1 Test");
    System.out.printf("%s\n", H2);
    BS.reset(); // lol I lied

    boolean verifOut;
    ByteBuffer tmpBuff0 = ByteBuffer.allocate((int) (BS.getLength() / 2));

    System.out.printf("Writing [0 - %d] bytes into w1_0_j.tmp\n", (int) (BS.getLength() / 2));
    FileOutputStream fcOut1_0 = new FileOutputStream("../files/w1_0_j.tmp");
    BS.write(fcOut1_0, 0, BS.getLength() / 2);
    fcOut1_0.close();

    System.out.printf("Checking integrity -> w1_0_j.tmp\n");
    FileChannel fcIn1_0 = FileChannel.open(FileSystems.getDefault()
        .getPath("../files/w1_0_j.tmp"), StandardOpenOption.READ);
    fcIn1_0.read(tmpBuff0, 0);
    verifOut = true;
    for (int i = 0; i < (int) (BS.getLength() / 2) && verifOut; ++i) {
      verifOut = BS.getByteStream()[i] == tmpBuff0.array()[i];
    }
    System.out.printf("Result: %s\n", verifOut ? "OK" : "Error");
    fcIn1_0.close();
    System.out.printf("%s\n", H2);
    System.out.printf("%s\n", H1);
    /* **************************************************** */
    System.out.printf("%36s\n", "write_0 Test");
    System.out.printf("%s\n", H2);
    BS.reset(); // lol I lied
    ByteBuffer tmpBuff2 = ByteBuffer.allocate((int) BS.getLength());

    System.out.printf("Writing everything into w0_j.tmp\n");
    FileOutputStream fcOut0 = new FileOutputStream("../files/w0_j.tmp");
    BS.write(fcOut0);
    fcOut0.close();

    System.out.printf("Checking integrity -> w0_j.tmp\n");
    FileChannel fcIn0 = FileChannel.open(FileSystems.getDefault()
        .getPath("../files/w0_j.tmp"), StandardOpenOption.READ);
    fcIn0.read(tmpBuff2, 0);
    verifOut = true;
    for (int i = 0; i < BS.getLength() && verifOut; ++i) {
      verifOut = BS.getByteStream()[i] == tmpBuff2.array()[i];
    }
    System.out.printf("Result: %s\n", verifOut ? "OK" : "Error");
    fcIn0.close();

    System.out.printf("%s\n", H1);
    /* **************************************************** */
  }
}

File: c/test/BSTest.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "../src/ByteStream.h"

#define H1 "**************************************************************************"
#define H2 "__________________________________________________________________________"

#define pack4B(a, b, c, d) *((int*) (signed char[4] ) {a, b, c, d} )

signed char lookUpByte_0(ByteStream *BS, int targetIndex) {
  for (int i = 0; i < targetIndex; ++i) {
    getByte_0(BS);
  }
  return getByte_0(BS);
}

int main() {
  printf("%s\n", H1);
  printf("Initializing ByteStream (normal mode)\n");

  ByteStream *BS = ByteStream_0();
  ByteBuffer arr = {
    (signed char []){0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15},
    16
  };
  signed char *ptr = BS->buffer.array;
  long long auxLen = BS->buffer.length;
  int verif;

  printf("%s\n", H1);
  printf("%41s\n", "putByte Test");
  printf("%s\n", H2);

  for (int i = 0; i < 10; i++) {
    ptr = BS->buffer.array;
    auxLen = BS->buffer.length;
    verif = 1;
    while (verif) {
      putByte(BS, 1); // [0 - 255], overflow not allowed
      verif = auxLen == BS->buffer.length;
    }
    printf("old: %p\tnew: %p\tlength: %lld\n",
      ptr, BS->buffer.array, BS->buffer.length);
  }
  printf("%s\n", H2);
  ptr = BS->buffer.array;
  printf("0: %d\t\t65536: %d\t\t524287: %d\t\t1000000: %d\n",
            ptr[0], ptr[65536], ptr[524287], ptr[1000000]);
  destroy(BS);
  free(BS);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%42s\n", "putBytes_0 Test");
  printf("%s\n", H2);
  BS = ByteStream_0();
  for (int i = 0; i < 10; i++) {
    ptr = BS->buffer.array;
    auxLen = BS->buffer.length;
    verif = 1;
    while (verif) {
      putBytes_0(BS, arr, 0, 16); // Whole array, sequences of [0-16]
      verif = auxLen == BS->buffer.length;
    }
    printf("old: %p\tnew: %p\tlength: %lld\n",
      ptr, BS->buffer.array, BS->buffer.length);
  }
  printf("%s\n", H2);
  ptr = BS->buffer.array;
  printf("0: %d\t\t67541: %d\t\t540639: %d\t\t1000000: %d\n",
            ptr[0], ptr[67541], ptr[540639], ptr[1000000]);
  destroy(BS);
  free(BS);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%42s\n", "putBytes_1 Test");
  printf("%s\n", H2);
  BS = ByteStream_0();
  for (int i = 0; i < 10; i++) {
    ptr = BS->buffer.array;
    auxLen = BS->buffer.length;
    verif = 1;
    while (verif) {
      putBytes_1(BS, pack4B(1, 63, 127, 255), 4); // Max 4B value
      verif = auxLen == BS->buffer.length;
    }
    printf("old: %p\tnew: %p\tlength: %lld\n",
      ptr, BS->buffer.array, BS->buffer.length);
  }
  printf("%s\n", H2);
  ptr = BS->buffer.array;
  printf("0: %d\t\t65534: %d\t\t524287: %d\t\t1000000: %d\n",
            ptr[0], ptr[65534], ptr[524287], ptr[1000000]);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "getByte_1 Test (normal mode)");
  printf("%s\n", H2);
  printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS, 0), ptr[0]);
  printf("%s\n", H2);
  printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS, 65534), ptr[65534]);
  printf("%s\n", H2);
  printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS, 524287), ptr[524287]);
  printf("%s\n", H2);
  printf("Index: limit-1\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS, BS->limit - 1), ptr[BS->limit - 1]);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "getByte_0 Test (normal mode)");
  printf("%s\n", H2);

  // Just don't want employ resets here for future tests
  printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS, 0), ptr[0]);
  printf("%s\n", H2);
  printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS, 65534 - 1), ptr[65534]);
  printf("%s\n", H2);
  printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS, 524287 - 65535), ptr[524287]);
  printf("%s\n", H2);
  printf("Index: limit-1\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS, BS->limit - 1 - 524288), ptr[BS->limit - 1]);
  printf("%s\n", H2);
  /* **************************************************** */
  printf("%36s\n", "getBytes Test (normal mode)");
  printf("%s\n", H2);
  ByteStream_reset(BS); // lol I lied

  printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", getBytes(BS, 1), pack4B(ptr[0], 0, 0, 0));
  printf("%s\n", H2);
  printf("Index: 1-2\t\t Got: %d\t\tExpected: %d\n", getBytes(BS, 2), pack4B(ptr[2], ptr[1], 0, 0));
  printf("%s\n", H2);
  printf("Index: 3-5\t\t Got: %d\t\tExpected: %d\n", getBytes(BS, 3), pack4B(ptr[5], ptr[4], ptr[3], 0));
  printf("%s\n", H2);
  printf("Index: 6-9\t\t Got: %d\t\tExpected: %d\n", getBytes(BS, 4), pack4B(ptr[9], ptr[8], ptr[7], ptr[6]));
  printf("%s\n", H2);
  printf("Index: 10-14\t\t Got: %d \t\tExpected: %d\n", getBytes(BS, 5),
    pack4B(ptr[14], ptr[13], ptr[12], ptr[11]) + pack4B(ptr[10], 0, 0, 0));
  printf("%s\n", H2);
  printf("Index: 15-22\t\t Got: %d \tExpected: %d\n", getBytes(BS, 8),
    pack4B(ptr[22], ptr[21], ptr[20], ptr[19]) + pack4B(ptr[18], ptr[17], ptr[16], ptr[15]));
  printf("%s\n", H2);
  /* Code for displaying the bits of the output number step by step in function getBytes(BS, num)

      // Simplified status of the buffer
      printf("\t\t%d\t%d|%d|%d|%d\n", (signed char) getByte, (num >> 0) & 0xff, (num >> 8) & 0xff, (num >> 16) & 0xff, (num >> 24) & 0xff);

      // New byte added
      printf("(%d) -> %d = ", (signed char) getByte, b * 8);
      for (int j = 7; j >= 0; j--) {
        printf("%c", (((signed char) getByte) & (1 << j)) ? '1' : '0');
      }
      printf(" ");
      printf("\n");

      // Detailed status of the buffer
      printf("(%d) %*s = ", num, 8, "->");
      for (int i = 0; i <= 3; i++) {
        for (int j = 7; j >= 0; j--) {
          printf("%c", ((num >> (i * 8)) & (1 << j)) ? '1' : '0');
        }
        printf(" ");
      }
      printf("\n");
  */
  /* **************************************************** */
  printf("%36s\n", "write_1 Test");
  printf("%s\n", H2);
  ByteStream_reset(BS); // lol I lied

  int verifOut;
  ByteBuffer tmpBuff0 = {(signed char *) malloc(getLength(BS) / 2), getLength(BS) / 2};

  printf("Writing [0 - %lld] bytes into w1_0_c.tmp\n", getLength(BS) / 2);
  FileChannel fcOut1_0 = FileChannel_0("../../files/w1_0_c.tmp", FC_WRITE);
  write_1(BS, fcOut1_0, 0, getLength(BS) / 2);
  fcClose(&fcOut1_0);

  printf("Checking integrity -> w1_0_c.tmp\n");
  FileChannel fcIn1_0 = FileChannel_0("../../files/w1_0_c.tmp", FC_READ);
  fcRead(&fcIn1_0, tmpBuff0, 0);
  verifOut = 1;
  for (long long i = 0; i < getLength(BS) / 2 && verifOut; ++i) {
    verifOut = ptr[i] == tmpBuff0.array[i];
  }
  printf("Result: %s\n", verifOut ? "OK" : "Error");
  fcClose(&fcIn1_0);
  free(tmpBuff0.array);
  printf("%s\n", H2);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "write_0 Test");
  printf("%s\n", H2);
  ByteStream_reset(BS); // lol I lied
  ByteBuffer tmpBuff2 = {(signed char *) malloc(getLength(BS)), getLength(BS)};

  printf("Writing everything into w0_c.tmp\n");
  FileChannel fcOut0 = FileChannel_0("../../files/w0_c.tmp", FC_WRITE);
  write_0(BS, fcOut0);

  printf("Checking integrity -> w0_c.tmp\n");
  FileChannel fcIn0 = FileChannel_0("../../files/w0_c.tmp", FC_READ);
  fcRead(&fcIn0, tmpBuff2, 0);
  verifOut = 1;
  for (long long i = 0; i < getLength(BS) && verifOut; ++i) {
    verifOut = ptr[i] == tmpBuff2.array[i];
  }
  printf("Result: %s\n", verifOut ? "OK" : "Error");
  fcClose(&fcIn0);
  free(tmpBuff2.array);

  fcClose(&fcOut0);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%s\n", H1);
  printf("Initializing ByteStream (readFile mode)\n");

  FileChannel FC = FileChannel_0("../../files/w0_c.tmp", FC_READ);
  ByteStream *BS2 = ByteStream_2(FC);

  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "putFileSegment Test");
  printf("%s\n", H2);

  printf("Putting two segments of size (limit / 2)\n");
  putFileSegment(BS2,                0L, BS->limit / 4);
  putFileSegment(BS2,     BS->limit / 4, BS->limit / 4);
  putFileSegment(BS2, 2 * BS->limit / 4, BS->limit / 4);
  putFileSegment(BS2, 3 * BS->limit / 4, BS->limit / 4);
  printf("Limit\t\tGot: %lld\t\t\tExpected: %lld\n", BS2->limit, BS->limit);
  printf("NumSegments\tGot: %d\t\t\t\tExpected: %d\n", BS2->readFileNumSegments, 4);

  printf("%s\n", H2);
  for (int i = 0; i < 4; ++i) {
    printf("Segment %d\tGot: %lld, %-16lld\tExpected: %lld, %lld\n", i,
        BS2->readFileSegments.begins[i], BS2->readFileSegments.lengths[i],
        i * BS->limit / 4, BS->limit / 4);
  }
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "getByte_1 Test (readFile mode)");
  printf("%s\n", H2);
  printf("Checking integrity of buffer with normalMode ByteStream\n");
  printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS2, 0), getByte_1(BS, 0));
  printf("%s\n", H2);
  printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS2, 65534), getByte_1(BS, 65534));
  printf("%s\n", H2);
  printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS2, 524287), getByte_1(BS, 524287));
  printf("%s\n", H2);
  printf("Index: limit-1\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS2, BS2->limit - 1), getByte_1(BS, BS->limit - 1));
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "getByte_0 Test (readFile mode)");
  printf("%s\n", H2);

  // Just don't want employ resets here for future tests
  printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS2, 0), ptr[0]);
  printf("%s\n", H2);
  printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS2, 2047 - 1), ptr[2047]);
  printf("%s\n", H2);
  printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS2, 65535 - 2048), ptr[65535]);
  printf("%s\n", H2);
  printf("Index: limit-1\t\t Got: %d\t\tExpected: %d\n", lookUpByte_0(BS2, BS2->limit - 1 - 65536), ptr[BS->limit - 1]);
  printf("%s\n", H2);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "endReadFileMode Test");
  printf("%s\n", H2);

  endReadFileMode(BS2);
  // From here, BS2 has streamMode = 0
  printf("Verifying both normalMode buffers\n");
  printf("%s\n", H2);
  printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", BS2->buffer.array[0], BS->buffer.array[0]);
  printf("%s\n", H2);
  printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", BS2->buffer.array[65534], BS->buffer.array[65534]);
  printf("%s\n", H2);
  printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", BS2->buffer.array[524287], BS->buffer.array[524287]);
  printf("%s\n", H2);
  printf("Index: limit-1\t\t Got: %d\t\tExpected: %d\n", BS2->buffer.array[BS2->limit - 1], BS->buffer.array[BS->limit - 1]);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "saveToTemporalFile Test");
  printf("%s\n", H2);

  printf("Writing to the temporalFile\n");
  saveToTemporalFile(BS2, "../../files/");
  printf("%s\n", H2);
  printf("Checking integrity -> w0_c.tmp vs %s\n", BS2->temporalFileName);
  FileChannel fcInTmp0 = FileChannel_0("../../files/w0_c.tmp", FC_READ);
  FileChannel fcInTmp1 = FileChannel_0(BS2->temporalFileName, FC_READ);
  ByteBuffer tmpBuffTmp0 = {(signed char *) malloc(fcSize(&fcInTmp0)), fcSize(&fcInTmp0)};
  ByteBuffer tmpBuffTmp1 = {(signed char *) malloc(fcSize(&fcInTmp1)), fcSize(&fcInTmp1)};
  fcRead(&fcInTmp0, tmpBuffTmp0, 0);
  fcRead(&fcInTmp1, tmpBuffTmp1, 0);
  verifOut = 1;
  for (int i = 0; i < getLength(BS) / 2 && verifOut; ++i) {
    verifOut = tmpBuffTmp0.array[i] == tmpBuffTmp1.array[i];
  }
  printf("Result: %s\n", verifOut ? "OK" : "Error");
  fcClose(&fcInTmp0);
  fcClose(&fcInTmp1);
  free(tmpBuffTmp0.array);
  free(tmpBuffTmp1.array);

  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "write_0 (temporalFile mode) Test");
  printf("%s\n", H2);
  FileChannel FCTMP = FileChannel_0("../../files/wt_c.tmp", FC_WRITE);
  write_0(BS2, FCTMP);
  fcClose(&FCTMP);
  printf("%s\n", H2);
  printf("Checking integrity -> wt_c.tmp vs %s\n", BS2->temporalFileName);
  FileChannel fcInTmp2 = FileChannel_0("../../files/wt_c.tmp", FC_READ);
  FileChannel fcInTmp3 = FileChannel_0(BS2->temporalFileName, FC_READ);
  ByteBuffer tmpBuffTmp2 = {(signed char *) malloc(fcSize(&fcInTmp3)), fcSize(&fcInTmp2)};
  ByteBuffer tmpBuffTmp3 = {(signed char *) malloc(fcSize(&fcInTmp2)), fcSize(&fcInTmp3)};
  fcRead(&fcInTmp2, tmpBuffTmp2, 0);
  fcRead(&fcInTmp3, tmpBuffTmp3, 0);
  verifOut = 1;
  for (int i = 0; i < getLength(BS2) / 2 && verifOut; ++i) {
    verifOut = tmpBuffTmp2.array[i] == tmpBuffTmp3.array[i];
  }
  printf("Result: %s\n", verifOut ? "OK" : "Error");
  fcClose(&fcInTmp2);
  fcClose(&fcInTmp3);
  free(tmpBuffTmp2.array);
  free(tmpBuffTmp3.array);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "loadFromTemporalFile Test");
  printf("%s\n", H2);

  loadFromTemporalFile(BS2);
  printf("Verifying both normalMode buffers\n");
  printf("%s\n", H2);
  printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", BS2->buffer.array[0], BS->buffer.array[0]);
  printf("%s\n", H2);
  printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", BS2->buffer.array[65534], BS->buffer.array[65534]);
  printf("%s\n", H2);
  printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", BS2->buffer.array[524287], BS->buffer.array[524287]);
  printf("%s\n", H2);
  printf("Index: limit-1\t\t Got: %d\t\tExpected: %d\n", BS2->buffer.array[BS2->limit - 1], BS->buffer.array[BS->limit - 1]);
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "destroyTemporalFile Test");
  printf("%s\n", H2);
  printf("Destroying temporalFile %s\n", BS2->temporalFileName);
  destroyTemporalFile(BS2);
  printf("%s\n", BS2->temporalFileName == NULL ? "OK" : "Error");
  printf("%s\n", H1);
  /* **************************************************** */
  printf("%36s\n", "returnReadFileMode Test");
  printf("%s\n", H2);
  returnReadFileMode(BS2);
  printf("Index: 0\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS2, 0), getByte_1(BS, 0));
  printf("%s\n", H2);
  printf("Index: 65534\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS2, 65534), getByte_1(BS, 65534));
  printf("%s\n", H2);
  printf("Index: 524287\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS2, 524287), getByte_1(BS, 524287));
  printf("%s\n", H2);
  printf("Index: limit-1\t\t Got: %d\t\tExpected: %d\n", getByte_1(BS2, BS2->limit - 1), getByte_1(BS, BS->limit - 1));
  printf("%s\n", H1);
  /* **************************************************** */
  destroy(BS);
  destroy(BS2);
  free(BS);
  free(BS2);
  /* **************************************************** */
  return(0);
}
